File: Source/Java/nl/devnology/domain/order/Order.java
package nl.devnology.domain.order;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import nl.devnology.domain.book.Book;
import nl.devnology.domain.customer.Customer;

/**
 * @author Erik Pragt
 */
public class Order {
    private Customer customer;

    private Map<Book, Integer> contents = new HashMap<Book, Integer>();

    public Order(Customer customer) {
        this.customer = customer;
    }

    public void addBook(Integer amount, Book book) {
        Integer currentAmount = contents.get(book);

        if(currentAmount == null) {
            currentAmount = 0;
        }

        currentAmount += amount;

        contents.put(book, currentAmount);
    }

    public Set<Book> getBooks() {
        return contents.keySet();
    }

    public Integer getAmount(Book book) {
        return contents.get(book);
    }

    public Customer getCustomer() {
        return customer;
    }

    public boolean containsBook(Book book) {
        return contents.keySet().contains(book);
    }

    public Integer getTotalPrice() {
        Integer totalPrice = 0;

        for (Map.Entry<Book, Integer> entry : contents.entrySet()) {
            totalPrice += (entry.getKey().getPrice() * entry.getValue());
        }
        return totalPrice;
    }
}

File: Source/C#/DevnologyFitnesseDojo/DevnologyFitnesseDojo/Domain/Order.cs

using System.Collections.Generic;

namespace DevnologyFitnesseDojo.Domain
{
    public class Order
    {
        private readonly Customer customer;
        private readonly Dictionary<Book, int> contents = new Dictionary<Book, int>();

        public Order(Customer customer)
        {
            this.customer = customer;
        }

        public void AddBook(int amount, Book book)
        {
            int currentAmount = 0;
            if (contents.ContainsKey(book))
            {
                currentAmount = contents[book];    
            }
            
            currentAmount += amount;

            contents[book] = currentAmount;
        }

        public IEnumerable<Book> Books
        {
            get { return contents.Keys; }
        }

        public int GetAmount(Book book)
        {
            return contents[book];
        }

        public Customer Customer
        {
            get
            {
                return customer;
            }
        }

        public bool ContainsBook(Book book)
        {
            return contents.ContainsKey(book);
        }

        public int TotalPrice
        {
            get
            {
                int totalPrice = 0;
                foreach (var content in contents)
                {
                    totalPrice += (int)content.Key.Price*content.Value;
                }
                return totalPrice;
            }
        }
    }
}
